zebra-state/src/service/check/tests/prop.rs

```diff
@@ -405,6 +920,111 @@ fn transaction_v4_with_joinsplit_data(
     }
 }
 
+/// Return a `Transaction::V4` containing `sapling_shielded_data`.
+/// with its `Spend`s replaced by `spends`.
+///
+/// Other fields have empty or default values.
+///
+/// Note: since sapling nullifiers in V5 transactions are identical to V4 transactions,
+/// we just use V4 transactions in the tests.
+///
+/// # Panics
+///
+/// If there are no `Spend`s in `spends`, and no `Output`s in `sapling_shielded_data`.
+fn transaction_v4_with_sapling_shielded_data<'until_cloned>(
+    sapling_shielded_data: impl Into<Option<sapling::ShieldedData<PerSpendAnchor>>>,
+    spends: impl IntoIterator<Item = &'until_cloned sapling::Spend<PerSpendAnchor>>,
+) -> Transaction {
+    let mut sapling_shielded_data = sapling_shielded_data.into();
+    let spends: Vec<_> = spends.into_iter().cloned().collect();
+
+    if let Some(ref mut sapling_shielded_data) = sapling_shielded_data {
+        // make sure there are no other nullifiers, by replacing all the spends
+        sapling_shielded_data.transfers = match (
+            sapling_shielded_data.transfers.clone(),
+            spends.is_empty(),
+        ) {
+            // old and new spends: replace spends
+            (
+                SpendsAndMaybeOutputs {
+                    shared_anchor,
+                    maybe_outputs,
+                    ..
+                },
+                false,
+            ) => SpendsAndMaybeOutputs {
+                shared_anchor,
+                spends: spends.try_into().expect("just checked at least one spend"),
+                maybe_outputs,
+            },
+            // old spends, but no new spends: delete spends, panic if no outputs
+            (SpendsAndMaybeOutputs { maybe_outputs, .. }, true) => JustOutputs {
+                outputs: maybe_outputs.try_into().expect(
+                    "unexpected invalid TransferData: must have at least one spend or one output",
+                ),
+            },
+            // no old spends, but new spends: add spends
+            (JustOutputs { outputs, .. }, false) => SpendsAndMaybeOutputs {
+                shared_anchor: FieldNotPresent,
+                spends: spends.try_into().expect("just checked at least one spend"),
+                maybe_outputs: outputs.into(),
+            },
+            // no old and no new spends: do nothing
+            (just_outputs @ JustOutputs { .. }, true) => just_outputs,
```

Comment:
Sapling spend replacement is complicated: there are 4 possible cases based on the structural rules and enum variants